x-pack/test/api_integration/apis/apm/agent_configuration.ts

```diff
@@ -219,6 +224,26 @@ export default function agentConfigurationTests({ getService }: FtrProviderConte
           return body._source.applied_by_agent;
         }
 
+        // wait until `applied_by_agent` has been updated in elasticsearch
+        expect(await waitFor(getAppliedByAgent)).to.be(true);
+      });
+      it(`should have 'applied_by_agent=false' before marking as applied`, async () => {
+        const res1 = await searchConfigurations({
+          service: { name: 'myservice', environment: 'production' },
+        });
+
+        expect(res1.body._source.applied_by_agent).to.be(false);
+      });
+      it(`should have 'applied_by_agent=true' when 'appliedByAgent' attribute is true`, async () => {
+        async function getAppliedByAgent() {
```

Comment:
I don't think you need a `waitFor` here, as it doesn't write to ES - in fact I'd say this test is slightly better without `waitFor`. Thoughts?

Reply:
Well, I can remove the `waitFor` but I believe I must wait for the result of `searchConfigurations`, which gets the configuration from ES. I don't know what would be the difference by doing:
```js
expect(await waitFor(getAppliedByAgent)).to.be(true);
```
or
```js
const appliedByAgent = await getAppliedByAgent();
expect(appliedByAgent).to.be(true);
```

Reply:
This is my understanding: `waitFor` has a retry function, because `markAppliedByAgent` is non-blocking, so if you immediately check for `applied_by_agent` after searching the first time you could run into a race condition where the write to ES (by `markAppliedByAgent`) has not been completed. When `applied_by_agent` is `true`, it should not matter if `markAppliedByAgent` from the previous search request has been resolved. Is that right?

Reply:
Yes, that's correct. It doesn't matter from which request the response came `true`. 
But I think it would make more sense to first fire a request which writes to ES then repeat only the search, like:
```js
it(`should have 'applied_by_agent=true' when 'appliedByAgent' attribute is true`, async () => {
        async function markAsAppliedByAgent() {
          const { body } = await searchConfigurations({
            service: { name: 'myservice', environment: 'production' },
            applied_by_agent: true,
          });

          return body._source.applied_by_agent;
        }

        async function getAppliedByAgent() {
          const { body } = await searchConfigurations({
            service: { name: 'myservice', environment: 'production' },
          });

          return body._source.applied_by_agent;
        }

        await markAsAppliedByAgent();

        // wait until `applied_by_agent` has been updated in elasticsearch
        expect(await waitFor(getAppliedByAgent)).to.be(true);
      });
```
The difference here is a request with `applied_by_agent: true` will write to ES and without it only search for the configuration.

Reply:
Ah, I misunderstood - I assumed that `applied_by_agent` would return rather than store `applied_by_agent` as `true`, but it writes similar to when etags are matched. That should be fine then and I think your suggestion is an improvement 👍

Also seems like `writeFor` is only retrying when the request fails, it doesn't wait for a condition to be true. Have we ever seen this test fail?

Reply:
Hmm, I'm not sure if I fully understand the use of `waitFor` here at all. Do you happen to know?

Reply:
The `waitFor` function, in this case, will keep fetching the configuration until `body._source.applied_by_agent` returns true, waiting 100ms between each call.

Reply:
Maybe we should change the name from `applied_by_agent` to `mark_as_applied_by_agent`, to avoid confusion. WDYT?

Reply:
Oh, I see now. This is a little hard to grok IMHO. I think we can improve a little further:

- use your proposed change as a base
- rename `getAppliedByAgent` to `hasBeenAppliedByAgent` 
- use `body._source.applied_by_agent === true` or `!!body._source.applied_by_agent`
- consistently use `waitFor` in all `applied_by_agent` tests

Reply:
I'm trying to think of how to appropriately use `waitFor` when checking if `applied_by_agent` is false, but I can't think of a way to do that without making a call to ES directly 🤔 that is probably not the responsibility of this test.

Reply:
@cauemarcondes I think that would be an improvement as well - you mean in the POST request to /search?

Reply:
yes, in the POST:/api/apm/settings/agent-configuration/search

Reply:
We don't use `waitFor` to check if `applied_by_agent` is false. Only when we expect that it should be true.

Reply:
I know, but that specific test (`should have 'applied_by_agent=false' before supplying etag`) could be a false positive because the write happens asynchronously. We'd need to have some kind of mechanism that checks whether the write has been completed and then do the second request. But we'd need to talk directly to ES in that case which I don't think this test should do.

Reply:
@simitt is it ok for you if we change the property name to `mark_as_applied_by_agent`?

Reply:
`should have 'applied_by_agent=false' before marking as applied` won't write anything to ES, since the `etag` and `mark_as_applied_by_agent` are missing in the request. It's only checking if the default value of `applied_by_agent` is false.

Reply:
yes, but I'm referring to `should have 'applied_by_agent=false' before supplying etag`. My point is: all requests could potentially write something to ES asynchronously, but we don't know when this is resolved and we should make a second request. The only workaround we have is retrying until `applied_by_agent` is true, and that's not ideal, but fine for now I think.